sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala

```diff
@@ -118,27 +119,26 @@ trait MultiFileReaderFunctions {
 // Singleton thread pool used across all tasks for multifile reading.
 // Please note that the TaskContext is not set in these threads and should not be used.
 object MultiFileReaderThreadPool extends Logging {
-  private var threadPool: Option[ThreadPoolExecutor] = None
 
-  private def initThreadPool(
-      numThreadsFromConf: Int,
-      keepAliveSeconds: Int = 60): ThreadPoolExecutor = synchronized {
-    if (threadPool.isEmpty) {
-      val numThreads = Math.max(numThreadsFromConf, GpuDeviceManager.getNumCores)
-
-      if (numThreadsFromConf != numThreads) {
-        logWarning(s"Configuring the file reader thread pool with a max of $numThreads " +
-            s"threads instead of ${RapidsConf.MULTITHREAD_READ_NUM_THREADS} = $numThreadsFromConf")
-      }
-
-      val threadPoolExecutor =
-        TrampolineUtil.newDaemonCachedThreadPool("multithreaded file reader worker", numThreads,
-          keepAliveSeconds)
-      threadPoolExecutor.allowCoreThreadTimeOut(true)
-      logDebug(s"Using $numThreads for the multithreaded reader thread pool")
-      threadPool = Some(threadPoolExecutor)
+  // Use ThreadLocal to keep thread specific. The singleton is guarded by the lazy variable.
+  private val numThreadsLocal: ThreadLocal[Int] = new ThreadLocal[Int]
+
+  private lazy val threadPool: ThreadPoolExecutor = {
```

Comment:
It's not a big deal for this PR but not sure it is a great use case for lazy val after all because we never touch it without passing numThreads via getOrCreateThreadPool. Not sure it is much better than just making getOrCreateThreadPool a synchronized method.